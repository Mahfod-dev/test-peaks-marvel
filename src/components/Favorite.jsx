import { Link } from 'react-router-dom';
import { getHeroImage } from '../helpers/getHeroById';

/**
 * Favorite is a reusable component that displays a card of favorites.
 * @component
 * @param {number} id - The id of the hero.
 * @param {string} name - The name of the hero.
 * @param {object} thumbnail - The thumbnail of the hero.
 * @param {func} onDeleteFavorite - The function to delete a favorite.
 * @returns {JSX.Element} - A JSX element that displays a card of favorites.
 * @example
 *
 * const id = 1;
 * const name = 'superman';
 * const thumbnail = { path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/50/537ba56d31087', extension: 'jpg' };
 * const onDeleteFavorite = () => {};
 * return (
 * <Favorite id={id} name={name} thumbnail={thumbnail} onDeleteFavorite={onDeleteFavorite} />
 * )
 *
 * */

export const Favorite = ({ id, name, thumbnail, onDeleteFavorite }) => {
	const image = getHeroImage(thumbnail);

	return (
		<div style={{ height: '100vh', marginTop: '10px' }}>
			<div className='card ms-3' style={{ maxWidth: 750 }}>
				<div className='row no-gutters'>
					<div className='col-md-4'>
						<img src={image} className='card-img' alt={name} />
					</div>
					<div className='col-md-8'>
						<div className='card-body d-flex row'>
							<Link to={`/hero/${id}`} className='card-title'>
								{name}
							</Link>
							<button
								className='btn btn-primary w-25'
								onClick={() => onDeleteFavorite(id)}>
								Delete Favori
							</button>
						</div>
					</div>
				</div>
			</div>
		</div>
	);
};
