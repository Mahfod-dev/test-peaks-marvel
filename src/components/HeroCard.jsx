import { Link } from "react-router-dom";
import { getHeroImage } from "../helpers/getHeroById";
import "../style/HeroesCard.css";
import PropTypes from "prop-types";

/**
 * HeroCard is a reusable component that displays a card of heroes.
 *
 * @component
 * @param {number} id - The id of the hero.
 * @param {string} name - The name of the hero.
 * @param {string} description - The description of the hero.
 * @param {object} thumbnail - The thumbnail of the hero.
 * @param {func} getHeroImage - The function to get the hero image.
 * @returns {JSX.Element} - A JSX element that displays a card of heroes.
 * @example
 * const id = 1;
 * const name='superman';
 * const description='superman is a hero';
 * const thumbnail = { path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/50/537ba56d31087', extension: 'jpg' };
 * return (
 *  <HeroCard id={id} name={name} description={description} thumbnail={thumbnail} />
 * )
 *
 *
 *
 */

export function HeroCard({ id, description, name, thumbnail }) {
  const heroImageUrl = getHeroImage(thumbnail);

  return (
    <>
      <div className="my-card">
        <img srcSet={heroImageUrl} alt={name} width="100%" height="100%" />

        <Link to={`/hero/${id}`} className="profile-name">
          {name}
        </Link>
        <div className="profile-overview">
          {description ? (
            <p className="card-text">{description}</p>
          ) : (
            <p className="card-text">sorry we will soon have the description</p>
          )}
        </div>
      </div>
    </>
  );
}

HeroCard.propTypes = {
  /**
   * The id of the hero.
   */
  id: PropTypes.number.isRequired,
  /**
   * The name of the hero.
   */
  name: PropTypes.string.isRequired,
  /**
   * The description of the hero.
   * @default 'sorry we will soon have the description'
   */
  description: PropTypes.string,
  /**
   * The thumbnail of the hero.
   * @default 'http://i.annihil.us/u/prod/marvel/i/mg/3/50/537ba56d31087'
   * @default 'jpg'
   */
  thumbnail: PropTypes.object.isRequired,
};
