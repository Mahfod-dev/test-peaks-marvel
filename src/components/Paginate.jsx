import Button from './ui/Button';

/**
 *  Paginate is a component that shows the pagination.
 * @param {object} heroes - The heroes.
 * @param {number} page - The page.
 * @param {function} nextPage - The next page function.
 * @param {function} prevPage - The previous page function.
 * @component
 * @example
 * return (
 * <Paginate />
 * )
 * @returns {JSX.Element} - A JSX element that displays a Paginate.

 * */

export const Paginate = ({ heroes, page, nextPage, prevPage }) => {
	return (
		<div
			style={{ display: 'flex', justifyContent: 'center', marginTop: '20px' }}>
			<nav aria-label='Page navigation'>
				<ul className='pagination'>
					<li className='page-item'>
						<Button
							onClick={prevPage}
							ariaLabel='Previous'
							description='&laquo;'
							name='Previous'
						/>
					</li>
					{heroes.map((_, index) => {
						return (
							<li key={index} className='page-item'>
								<button
									data-testid='paginate-button'
									className={`page-link ${index === page ? 'active' : null}`}>
									{index + 1}
								</button>
							</li>
						);
					})}

					<li className='page-item'>
						<Button
							onClick={nextPage}
							ariaLabel='Next'
							description='&raquo;'
							name='Next'
						/>
					</li>
				</ul>
			</nav>
		</div>
	);
};
