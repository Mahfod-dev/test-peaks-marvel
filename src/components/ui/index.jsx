export * from './Navbar';
export * from './NoFavorite';
export * from './Spinner';
export * from './Button';
export * from './Footer';
