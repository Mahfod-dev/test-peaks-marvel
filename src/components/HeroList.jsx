import { useHeroes } from '../context/HeroesContext';
import { HeroCard } from './HeroCard';

export const HeroList = () => {
	const { heroes } = useHeroes();
	
	return (
		<div className='row row-cols-1 row-cols-md-3 g-3'>
			{heroes.map((hero) => {
				return <HeroCard key={hero.id} {...hero} />;
			})}
		</div>
	);
};
