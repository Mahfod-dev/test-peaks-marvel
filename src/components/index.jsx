import Navbar from './ui/Navbar';
import HeroList from './HeroList';
import HeroCard from './HeroCard';

export { Navbar, HeroList, HeroCard };
