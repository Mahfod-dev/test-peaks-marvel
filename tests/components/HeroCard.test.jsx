import { describe, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import { HeroCard } from '../../src/components/HeroCard';
import { BrowserRouter } from 'react-router-dom';

describe('HeroCard', () => {
	const id = 1017100;
	const name = 'A-Bomb (HAS)';
	const description =
		"Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction!  A-Bomb (HAS)";
	const thumbnail = {
		path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16',
		extension: 'jpg',
	};

	test('should render match snapshot ', () => {
		const { container } = render(
			<HeroCard
				id={id}
				name={name}
				description={description}
				thumbnail={thumbnail}
			/>,
			{ wrapper: BrowserRouter }
		);

		expect(container).toMatchSnapshot();
	});

	test('should render image with src and alt', () => {
		render(
			<HeroCard
				id={id}
				name={name}
				description={description}
				thumbnail={thumbnail}
			/>,
			{ wrapper: BrowserRouter }
		);

		

		const { src, alt } = screen.getByRole('img');

		expect(src).toBe(thumbnail.path + '.' + thumbnail.extension);
		expect(alt).toBe(name);
	});
});
