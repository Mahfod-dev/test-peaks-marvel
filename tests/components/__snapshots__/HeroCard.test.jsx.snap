// Vitest Snapshot v1

exports[`HeroCard > should render match snapshot  1`] = `
<div>
  <div
    class="my-card"
  >
    <img
      alt="A-Bomb (HAS)"
      srcset="http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16.jpg"
      style="width: 615px; height: 350px;"
    />
    <a
      class="profile-name"
      href="/hero/1017100"
    >
      A-Bomb (HAS)
    </a>
    <div
      class="profile-overview"
    >
      <p
        class="card-text"
      >
        Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction!  A-Bomb (HAS)
      </p>
    </div>
  </div>
</div>
`;
